From 8e0f99eff5d8462c552abc83bc347bcc9d868cfd Mon Sep 17 00:00:00 2001
From: Juliusz Sosinowicz <juliusz@wolfssl.com>
Date: Fri, 4 Sep 2026 11:17:54 +0000
Subject: [PATCH] Add wolfSSL support to OpenLDAP 2.6.13

Adds --with-tls=wolfssl to configure.ac and ports libldap/slapd to the
wolfSSL OpenSSL compatibility layer.

Compile wolfSSL with:
  ./configure --enable-openldap CPPFLAGS=-DWOLFSSL_NO_ASN_STRICT
  make
  make install

WOLFSSL_NO_ASN_STRICT is needed for the test suite only: the test
certificates in tests/data use a serial number of 0.

TLSCRLCheck needs CRL support in wolfSSL. --enable-openldap turns it on
(older wolfSSL releases need --enable-crl as well). Without it libldap
refuses the option instead of silently ignoring it.

Compile OpenLDAP with:
  patch -p1 < openldap-2.6.13.patch
  rm aclocal.m4
  autoreconf -ivf
  ./configure --with-tls=wolfssl --disable-bdb --disable-hdb \
    CFLAGS="-I/usr/local/include -I/usr/local/include/wolfssl \
      -L/usr/local/lib"
  make depend
  make

autoreconf may complain at the automake step because OpenLDAP has no
Makefile.am. That is expected; autoheader and autoconf still run.

Run the tests with:
  make check

All 90 tests in tests/scripts run for the mdb backend: 40 pass and 50
are skipped by the test suite itself because the corresponding
overlays, backends (ldap, meta, asyncmeta), SASL, ACI or external
services (MSAD, DSEE) are not built or configured with the configure
line above. The same 50 tests are skipped when the same tree is built
against OpenSSL, so the skips are not caused by wolfSSL. The TLS test
test067-tls (StartTLS, ldaps, hard cert validation and key pinning)
and the certificate test test021-certificate pass. Upstream runs the
tls_protocol_min sub-check of test067-tls only for OpenSSL.
---
 configure.ac              |  40 +++++++++++-
 libraries/libldap/init.c  |   2 +-
 libraries/libldap/tls2.c  |  17 +++--
 libraries/libldap/tls_o.c | 130 ++++++++++++++++++++++++++++++++++----
 servers/lloadd/config.c   |   6 +-
 servers/slapd/bconfig.c   |   2 +-
 servers/slapd/config.c    |   8 +--
 servers/slapd/slap.h      |   2 +-
 8 files changed, 177 insertions(+), 30 deletions(-)

diff --git a/configure.ac b/configure.ac
index 27377ac..dd3287b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -247,8 +247,8 @@ OL_ARG_WITH(threads,
 	[AS_HELP_STRING([--with-threads], [with threads library auto|nt|posix|pth|lwp|manual])],
 	auto, [auto nt posix pth lwp yes no manual] )
 OL_ARG_WITH(tls,
-	[AS_HELP_STRING([--with-tls], [with TLS/SSL support auto|openssl|gnutls])],
-	auto, [auto openssl gnutls yes no] )
+	[AS_HELP_STRING([--with-tls], [with TLS/SSL support auto|openssl|gnutls|wolfssl])],
+	auto, [auto openssl gnutls wolfssl yes no] )
 OL_ARG_WITH(yielding_select,
 	[AS_HELP_STRING([--with-yielding-select], [with implicitly yielding select])],
 	auto, [auto yes no manual] )
@@ -1293,6 +1293,42 @@ if test $ol_link_tls = no ; then
 	fi
 fi
 
+dnl NOTE: caller must specify CPPFLAGS=-I/path/to/wolfssl/ to pick up
+dnl OpenSSL compatibility headers (e.g. /usr/local/include/wolfssl).
+if test $ol_link_tls = no ; then
+       if test $ol_with_tls = wolfssl || test $ol_with_tls = auto ; then
+               have_wolfssl=no
+
+               AC_CHECK_HEADERS([wolfssl/options.h])
+               if test "$ac_cv_header_wolfssl_options_h" = no &&
+                  test $ol_with_tls = wolfssl ; then
+                       AC_MSG_ERROR([wolfSSL not found])
+               fi
+
+               AC_RUN_IFELSE([AC_LANG_PROGRAM([
+               #include <wolfssl/options.h>
+               ],[
+               #ifndef OPENSSL_EXTRA
+               exit(1);
+               #endif
+               ])], [have_wolfssl=yes])
+
+               if test "$have_wolfssl" = yes ; then
+                       ol_with_tls=wolfssl
+                       ol_link_tls=yes
+
+                       TLS_LIBS="-lwolfssl"
+
+                       AC_DEFINE(HAVE_WOLFSSL, 1,
+                               [define if you have wolfSSL])
+               else
+                       if test $ol_with_tls = wolfssl ; then
+                               AC_MSG_ERROR([wolfSSL not compiled with OpenSSL compatibility layer. Re-compile and install wolfSSL with OPENSSL_EXTRA defined.])
+                       fi
+               fi
+       fi
+fi
+
 WITH_TLS=no
 if test $ol_link_tls = yes ; then
 	AC_DEFINE(HAVE_TLS, 1, [define if you have TLS])
diff --git a/libraries/libldap/init.c b/libraries/libldap/init.c
index e40fa93..5692641 100644
--- a/libraries/libldap/init.c
+++ b/libraries/libldap/init.c
@@ -133,7 +133,7 @@ static const struct ol_attribute {
 	{0, ATTR_TLS,	"TLS_PEERKEY_HASH",	NULL,	LDAP_OPT_X_TLS_PEERKEY_HASH},
 	{0, ATTR_TLS,	"TLS_ECNAME",		NULL,	LDAP_OPT_X_TLS_ECNAME},
 
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	{0, ATTR_TLS,	"TLS_CRLCHECK",		NULL,	LDAP_OPT_X_TLS_CRLCHECK},
 #endif
 #ifdef HAVE_GNUTLS
diff --git a/libraries/libldap/tls2.c b/libraries/libldap/tls2.c
index 054b016..f85d8ca 100644
--- a/libraries/libldap/tls2.c
+++ b/libraries/libldap/tls2.c
@@ -551,6 +551,10 @@ ldap_pvt_tls_get_peer_dn( void *s, struct berval *dn,
 
 	rc = ldap_X509dn2bv( &bvdn, dn, 
 			    (LDAPDN_rewrite_func *)func, flags);
+#ifdef HAVE_WOLFSSL
+	/* see comment in tlso_session_peer_dn */
+	free(bvdn.bv_val);
+#endif
 	return rc;
 }
 
@@ -646,7 +650,7 @@ ldap_pvt_tls_config( LDAP *ld, int option, const char *arg )
 		}
 		return ldap_pvt_tls_set_option( ld, option, &i );
 		}
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	case LDAP_OPT_X_TLS_CRLCHECK:	/* OpenSSL only */
 		i = -1;
 		if ( strcasecmp( arg, "none" ) == 0 ) {
@@ -736,7 +740,7 @@ ldap_pvt_tls_get_option( LDAP *ld, int option, void *arg )
 	case LDAP_OPT_X_TLS_REQUIRE_SAN:
 		*(int *)arg = lo->ldo_tls_require_san;
 		break;
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	case LDAP_OPT_X_TLS_CRLCHECK:	/* OpenSSL only */
 		*(int *)arg = lo->ldo_tls_crlcheck;
 		break;
@@ -957,7 +961,7 @@ ldap_pvt_tls_set_option( LDAP *ld, int option, void *arg )
 			return 0;
 		}
 		return -1;
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	case LDAP_OPT_X_TLS_CRLCHECK:	/* OpenSSL only */
 		if ( !arg ) return -1;
 		switch( *(int *) arg ) {
@@ -1267,8 +1271,13 @@ ldap_pvt_tls_get_my_dn( void *s, struct berval *dn, LDAPDN_rewrite_dummy *func,
 	int rc;
 
 	rc = tls_imp->ti_session_my_dn( session, &der_dn );
-	if ( rc == LDAP_SUCCESS )
+	if ( rc == LDAP_SUCCESS ) {
 		rc = ldap_X509dn2bv(&der_dn, dn, (LDAPDN_rewrite_func *)func, flags );
+#ifdef HAVE_WOLFSSL
+		/* see comment in tlso_session_peer_dn */
+		free(der_dn.bv_val);
+#endif
+	}
 	return rc;
 }
 
diff --git a/libraries/libldap/tls_o.c b/libraries/libldap/tls_o.c
index cad42f8..872c4be 100644
--- a/libraries/libldap/tls_o.c
+++ b/libraries/libldap/tls_o.c
@@ -18,7 +18,7 @@
 
 #include "portable.h"
 
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 
 #include "ldap_config.h"
 
@@ -37,18 +37,23 @@
 #include "ldap-int.h"
 #include "ldap-tls.h"
 
-#ifdef HAVE_OPENSSL_SSL_H
+#if defined(HAVE_OPENSSL_SSL_H) || defined(HAVE_WOLFSSL)
+#ifdef HAVE_WOLFSSL
+#include <wolfssl/options.h>
+#endif
 #include <openssl/ssl.h>
 #include <openssl/x509v3.h>
 #include <openssl/err.h>
 #include <openssl/rand.h>
+#ifndef HAVE_WOLFSSL
 #include <openssl/safestack.h>
+#endif
 #include <openssl/bn.h>
 #include <openssl/rsa.h>
 #include <openssl/dh.h>
 #endif
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000
+#if OPENSSL_VERSION_NUMBER >= 0x10100000 && !defined(HAVE_WOLFSSL)
 #define ASN1_STRING_data(x)	ASN1_STRING_get0_data(x)
 #endif
 
@@ -93,7 +98,7 @@ static void tlso_locking_cb( int mode, int type, const char *file, int line )
 	}
 }
 
-#if OPENSSL_VERSION_NUMBER >= 0x0909000
+#if OPENSSL_VERSION_NUMBER >= 0x0909000 && !defined(HAVE_WOLFSSL)
 static void tlso_thread_self( CRYPTO_THREADID *id )
 {
 	CRYPTO_THREADID_set_pointer( id, (void *)ldap_pvt_thread_self() );
@@ -131,7 +136,7 @@ static void tlso_thr_init( void ) {}
 #endif
 #endif /* OpenSSL 1.1 */
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000
+#if OPENSSL_VERSION_NUMBER < 0x10100000 && !defined(HAVE_WOLFSSL)
 /*
  * OpenSSL 1.1 API and later makes the BIO method concrete types internal.
  */
@@ -168,6 +173,7 @@ BIO_meth_free( BIO_METHOD *meth )
 
 #endif /* OpenSSL 1.1 */
 
+#ifndef HAVE_WOLFSSL
 static STACK_OF(X509_NAME) *
 tlso_ca_list( char * bundle, char * dir, X509 *cert )
 {
@@ -203,6 +209,7 @@ tlso_ca_list( char * bundle, char * dir, X509 *cert )
 	}
 	return ca_list;
 }
+#endif
 
 /*
  * Initialize TLS subsystem. Should be called only once.
@@ -235,8 +242,10 @@ tlso_init( void )
 	OPENSSL_init_ssl(OPENSSL_FLAGS, NULL);
 #endif
 
+#ifndef HAVE_WOLFSSL
 	/* FIXME: mod_ssl does this */
 	X509V3_add_standard_extensions();
+#endif
 
 	tlso_bio_method = tlso_bio_setup();
 
@@ -275,7 +284,7 @@ static void
 tlso_ctx_ref( tls_ctx *ctx )
 {
 	tlso_ctx *c = (tlso_ctx *)ctx;
-#if OPENSSL_VERSION_NUMBER < 0x10100000
+#if OPENSSL_VERSION_NUMBER < 0x10100000 && !defined(HAVE_WOLFSSL)
 #define	SSL_CTX_up_ref(ctx)	CRYPTO_add( &(ctx->references), 1, CRYPTO_LOCK_SSL_CTX )
 #endif
 	SSL_CTX_up_ref( c );
@@ -511,8 +520,62 @@ tlso_ctx_init( struct ldapoptions *lo, struct ldaptls *lt, int is_server, char *
 
 		if ( is_server ) {
 			STACK_OF(X509_NAME) *calist;
+/*
+ * tlso_ca_list calls SSL_add_dir_cert_subjects_to_stack, which wolfSSL hasn't
+ * implemented. However, since we've just added all the CA certs in
+ * lt->lt_cacertdir to the wolfSSL cert manager, we can just get them from
+ * there, which is what this #ifdef'd code does instead of tlso_ca_list.
+ */
+#ifdef HAVE_WOLFSSL
+			WOLFSSL_CERT_MANAGER* cert_manager = NULL;
+			STACK_OF(X509)* certs = NULL;
+			X509* x509;
+			X509_NAME* subj_name;
+			int i;
+
+			if ((calist = sk_X509_NAME_new_null()) == NULL) {
+				Debug0(LDAP_DEBUG_ANY, "TLS: could not create X509_NAME stack\n");
+                tlso_report_error( errmsg );
+				return -1;
+			}
+			if ((cert_manager = wolfSSL_CTX_GetCertManager(ctx)) == NULL) {
+				Debug0(LDAP_DEBUG_ANY, "TLS: could not get cert manager\n");
+				sk_X509_NAME_pop_free(calist, NULL);
+                tlso_report_error( errmsg );
+				return -1;
+			}
+
+			if ((certs = wolfSSL_CertManagerGetCerts(cert_manager)) == NULL) {
+				Debug0(LDAP_DEBUG_ANY, "TLS: could not load certs from cert "
+				      "manager\n");
+				sk_X509_NAME_pop_free(calist, NULL);
+                tlso_report_error( errmsg );
+				return -1;
+			}
+
+			for (i = 0; i < sk_num(certs); ++i) {
+				x509 = sk_X509_value(certs, i);
+				if (X509_check_ca(x509)) {
+					subj_name = X509_NAME_dup(X509_get_subject_name(x509));
+					if (subj_name == NULL) {
+						Debug0(LDAP_DEBUG_ANY, "TLS: could not duplicate subject"
+						      " name\n");
+						sk_X509_NAME_pop_free(calist, NULL);
+						sk_X509_pop_free(certs, NULL);
+		                tlso_report_error( errmsg );
+						return -1;
+					}
+					else {
+						sk_X509_NAME_push(calist, subj_name);
+					}
+				}
+			}
+
+			sk_X509_pop_free(certs, NULL);
+#else
 			/* List of CA names to send to a client */
 			calist = tlso_ca_list( lt->lt_cacertfile, lt->lt_cacertdir, cert );
+#endif /* HAVE_WOLFSSL */
 			if ( !calist ) {
 				Debug2( LDAP_DEBUG_ANY, "TLS: "
 					"could not load client CA list (file:`%s',dir:`%s').\n",
@@ -627,7 +690,7 @@ tlso_ctx_init( struct ldapoptions *lo, struct ldaptls *lt, int is_server, char *
 	 * This is a NOP in OpenSSL 1.1.0 and later, where curves are always
 	 * auto-negotiated.
 	 */
-#if OPENSSL_VERSION_NUMBER < 0x10100000UL
+#if OPENSSL_VERSION_NUMBER < 0x10100000UL && !defined(HAVE_WOLFSSL)
 		if ( SSL_CTX_set_ecdh_auto( ctx, 1 ) <= 0 ) {
 			Debug0( LDAP_DEBUG_ANY,
 				"TLS: could not enable automatic EC negotiation.\n" );
@@ -657,12 +720,25 @@ tlso_ctx_init( struct ldapoptions *lo, struct ldaptls *lt, int is_server, char *
 #endif
 	if ( lo->ldo_tls_crlcheck ) {
 		X509_STORE *x509_s = SSL_CTX_get_cert_store( ctx );
+		int crc = 1;
 		if ( lo->ldo_tls_crlcheck == LDAP_OPT_X_TLS_CRL_PEER ) {
-			X509_STORE_set_flags( x509_s, X509_V_FLAG_CRL_CHECK );
+			crc = X509_STORE_set_flags( x509_s, X509_V_FLAG_CRL_CHECK );
 		} else if ( lo->ldo_tls_crlcheck == LDAP_OPT_X_TLS_CRL_ALL ) {
-			X509_STORE_set_flags( x509_s, 
+			crc = X509_STORE_set_flags( x509_s, 
 					X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL  );
 		}
+#ifdef HAVE_WOLFSSL
+		/* Without CRL support wolfSSL accepts the flag and ignores it. */
+		if ( crc != 1 ) {
+			Debug0( LDAP_DEBUG_ANY, "TLS: CRL checking requested but "
+				"wolfSSL was built without CRL support.\n" );
+			strncpy( errmsg, "wolfSSL was built without CRL support",
+				ERRBUFSIZE );
+			return -1;
+		}
+#else
+		(void) crc;
+#endif
 	}
 	/* Explicitly honor the server side cipher suite preference */
 	SSL_CTX_set_options( ctx, SSL_OP_CIPHER_SERVER_PREFERENCE );
@@ -752,13 +828,25 @@ tlso_session_my_dn( tls_session *sess, struct berval *der_dn )
 	tlso_session *s = (tlso_session *)sess;
 	X509 *x;
 	X509_NAME *xn;
+#ifdef HAVE_WOLFSSL
+	int ret;
+	unsigned char *der_buf = NULL;
+#endif
 
 	x = SSL_get_certificate( s );
 
 	if (!x) return LDAP_INVALID_CREDENTIALS;
 	
 	xn = X509_get_subject_name(x);
-#if OPENSSL_VERSION_NUMBER < 0x10100000
+#ifdef HAVE_WOLFSSL
+	/* der_buf is dynamically allocated, so callers of this function will need
+	   to free der_dn->bv_val. */
+	ret = i2d_X509_NAME(xn, &der_buf);
+	if (ret < 0)
+		return LDAP_INVALID_CREDENTIALS;
+	der_dn->bv_len = ret;
+	der_dn->bv_val = der_buf;
+#elif OPENSSL_VERSION_NUMBER < 0x10100000
 	der_dn->bv_len = i2d_X509_NAME( xn, NULL );
 	der_dn->bv_val = xn->bytes->data;
 #else
@@ -789,12 +877,24 @@ tlso_session_peer_dn( tls_session *sess, struct berval *der_dn )
 	tlso_session *s = (tlso_session *)sess;
 	X509 *x = tlso_get_cert( s );
 	X509_NAME *xn;
+#ifdef HAVE_WOLFSSL
+	unsigned char *der_buf = NULL;
+	int ret;
+#endif
 
 	if ( !x )
 		return LDAP_INVALID_CREDENTIALS;
 
 	xn = X509_get_subject_name(x);
-#if OPENSSL_VERSION_NUMBER < 0x10100000
+#ifdef HAVE_WOLFSSL
+	/* der_buf is dynamically allocated, so callers of this function will need
+	   to free der_dn->bv_val. */
+	ret = i2d_X509_NAME(xn, &der_buf);
+	if (ret < 0)
+		return LDAP_INVALID_CREDENTIALS;
+	der_dn->bv_len = ret;
+	der_dn->bv_val = der_buf;
+#elif OPENSSL_VERSION_NUMBER < 0x10100000
 	der_dn->bv_len = i2d_X509_NAME( xn, NULL );
 	der_dn->bv_val = xn->bytes->data;
 #else
@@ -1077,7 +1177,7 @@ tlso_session_endpoint( tls_session *sess, struct berval *buf, int is_server )
 	if ( cert == NULL )
 		return 0;
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000
+#if OPENSSL_VERSION_NUMBER >= 0x10100000 || defined(HAVE_WOLFSSL)
 	{
 		int mdnid;
 		if ( !OBJ_find_sigid_algs( X509_get_signature_nid( cert ), &mdnid, NULL ))
@@ -1090,6 +1190,7 @@ tlso_session_endpoint( tls_session *sess, struct berval *buf, int is_server )
 
 	/* See RFC 5929 */
 	if ( md == NULL ||
+#ifndef HAVE_WOLFSSL
 	     md == EVP_md_null() ||
 #ifndef OPENSSL_NO_MD2
 	     md == EVP_md2() ||
@@ -1097,6 +1198,7 @@ tlso_session_endpoint( tls_session *sess, struct berval *buf, int is_server )
 #ifndef OPENSSL_NO_MD4
 	     md == EVP_md4() ||
 #endif
+#endif
 #ifndef OPENSSL_NO_MD5
 	     md == EVP_md5() ||
 #endif
@@ -1237,7 +1339,7 @@ struct tls_data {
 	Sockbuf_IO_Desc		*sbiod;
 };
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000
+#if OPENSSL_VERSION_NUMBER < 0x10100000 && !defined(HAVE_WOLFSSL)
 #define BIO_set_init(b, x)	b->init = x
 #define BIO_set_data(b, x)	b->ptr = x
 #define BIO_clear_flags(b, x)	b->flags &= ~(x)
@@ -1759,4 +1861,4 @@ tls_impl ldap_int_tls_impl = {
 	0
 };
 
-#endif /* HAVE_OPENSSL */
+#endif /* HAVE_OPENSSL || HAVE_WOLFSSL */
diff --git a/servers/lloadd/config.c b/servers/lloadd/config.c
index d0e938b..b5ec152 100644
--- a/servers/lloadd/config.c
+++ b/servers/lloadd/config.c
@@ -511,7 +511,7 @@ static ConfigTable config_back_cf_table[] = {
         NULL, NULL
     },
     { "TLSCRLCheck", NULL, 2, 2, 0,
-#if defined(HAVE_TLS) && defined(HAVE_OPENSSL)
+#if defined(HAVE_TLS) && (defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL))
         CFG_TLS_CRLCHECK|ARG_STRING|ARG_MAGIC,
         &config_tls_config,
 #else
@@ -2868,7 +2868,7 @@ static slap_cf_aux_table bindkey[] = {
     { BER_BVC("tls_cipher_suite="), offsetof(slap_bindconf, sb_tls_cipher_suite), 's', 0, NULL },
     { BER_BVC("tls_protocol_min="), offsetof(slap_bindconf, sb_tls_protocol_min), 's', 0, NULL },
     { BER_BVC("tls_ecname="), offsetof(slap_bindconf, sb_tls_ecname), 's', 0, NULL },
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
     { BER_BVC("tls_crlcheck="), offsetof(slap_bindconf, sb_tls_crlcheck), 's', 0, NULL },
 #endif
 #endif
@@ -3220,7 +3220,7 @@ lload_bindconf_tls_set( slap_bindconf *bc, LDAP *ld )
             } else
                 newctx = 1;
         }
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
         if ( bc->sb_tls_crlcheck ) {
             rc = ldap_pvt_tls_config(
                     ld, LDAP_OPT_X_TLS_CRLCHECK, bc->sb_tls_crlcheck );
diff --git a/servers/slapd/bconfig.c b/servers/slapd/bconfig.c
index 9757ee6..42bc0ed 100644
--- a/servers/slapd/bconfig.c
+++ b/servers/slapd/bconfig.c
@@ -861,7 +861,7 @@ static ConfigTable config_back_cf_table[] = {
 			"EQUALITY caseExactMatch "
 			"SYNTAX OMsDirectoryString SINGLE-VALUE )", NULL, NULL },
 	{ "TLSCRLCheck", NULL, 2, 2, 0,
-#if defined(HAVE_TLS) && defined(HAVE_OPENSSL)
+#if defined(HAVE_TLS) && (defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL))
 		CFG_TLS_CRLCHECK|ARG_STRING|ARG_MAGIC, &config_tls_config,
 #else
 		ARG_IGNORED, NULL,
diff --git a/servers/slapd/config.c b/servers/slapd/config.c
index aa5aa6b..b7241c3 100644
--- a/servers/slapd/config.c
+++ b/servers/slapd/config.c
@@ -1295,7 +1295,7 @@ static slap_cf_aux_table bindkey[] = {
 	{ BER_BVC("tls_cipher_suite="), offsetof(slap_bindconf, sb_tls_cipher_suite), 's', 0, NULL },
 	{ BER_BVC("tls_protocol_min="), offsetof(slap_bindconf, sb_tls_protocol_min), 's', 0, NULL },
 	{ BER_BVC("tls_ecname="), offsetof(slap_bindconf, sb_tls_ecname), 's', 0, NULL },
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	{ BER_BVC("tls_crlcheck="), offsetof(slap_bindconf, sb_tls_crlcheck), 's', 0, NULL },
 #endif
 #endif
@@ -1676,7 +1676,7 @@ void bindconf_free( slap_bindconf *bc ) {
 		ch_free( bc->sb_tls_ecname );
 		bc->sb_tls_ecname = NULL;
 	}
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	if ( bc->sb_tls_crlcheck ) {
 		ch_free( bc->sb_tls_crlcheck );
 		bc->sb_tls_crlcheck = NULL;
@@ -1716,7 +1716,7 @@ bindconf_tls_defaults( slap_bindconf *bc )
 		if ( !bc->sb_tls_ecname )
 			slap_tls_get_config( slap_tls_ld, LDAP_OPT_X_TLS_ECNAME,
 				&bc->sb_tls_ecname );
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 		if ( !bc->sb_tls_crlcheck )
 			slap_tls_get_config( slap_tls_ld, LDAP_OPT_X_TLS_CRLCHECK,
 				&bc->sb_tls_crlcheck );
@@ -1798,7 +1798,7 @@ int bindconf_tls_set( slap_bindconf *bc, LDAP *ld )
 			} else
 				newctx = 1;
 		}
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 		if ( bc->sb_tls_crlcheck ) {
 			rc = ldap_pvt_tls_config( ld, LDAP_OPT_X_TLS_CRLCHECK,
 				bc->sb_tls_crlcheck );
diff --git a/servers/slapd/slap.h b/servers/slapd/slap.h
index e9b6ba7..554dc96 100644
--- a/servers/slapd/slap.h
+++ b/servers/slapd/slap.h
@@ -1668,7 +1668,7 @@ typedef struct slap_bindconf {
 	char *sb_tls_cipher_suite;
 	char *sb_tls_protocol_min;
 	char *sb_tls_ecname;
-#ifdef HAVE_OPENSSL
+#if defined(HAVE_OPENSSL) || defined(HAVE_WOLFSSL)
 	char *sb_tls_crlcheck;
 #endif
 	int sb_tls_int_reqcert;
-- 
2.43.0

